### PR TITLE
Allow AnimController to support any target

### PR DESCRIPTION
**Pull request by slimbuck**
The AnimController was written specifically to support GLB animations. However we would like to leverage the system toward controlling arbitrary targets in PlayCanvas.

This PR includes the following changes:
- introduce a general concept of animation target 'path' strings which specify what is being animated
- provide a method of escaping the target path segments so we can safely combine arbitrary path segments
- extract the code which 'binds' paths to target elements from the AnimController - this is called the AnimBinder
- allow the binder to provide the function which actually sets the resulting animation value on the target (since targets can be arbitrary)
- update the engine GLB viewer (the graphing visualization made use of internal AnimController structures before).
- update the animation component to correct set clip's speed and playing flag.

A summary of the anim API:
```javascript
// Interpolation
pc.INTERPOLATION_STEP       // step between curve values
pc.INTERPOLATION_LINEAR     // linearly interpolate between curve values
pc.INTERPOLATION_CUBIC      // perform cubic interpolation between curve values

// AnimData
animData = pc.AnimData(components, data:number[]);
animData.components     // number of values per element (1 for 1d values, 3 for 3d values etc)
animData.data           // list of numbers or Float32Array

// Curve
animCurve = pc.AnimCurve(paths:string[], key/input:number, value:number, interoplationType);
animCurve.paths         // the paths that will be targetted by this curve data
animCurve.input         // the input array
animCurve.value         // the output array

// Track
animTrack = pc.AnimTrack(name, duration, inputs:AnimData[], values:AnimData[], curves:AnimCurve[]);
animTrack.name          // name of the track
animTrack.duration      // duration in seconds
animTrack.inputs        // list of input (time/key) arrays
animTrack.outputs       // list of output (value) arrays
animTrack.curves        // list of curves

// Clip
animClip = pc.AnimClip(track, time, speed, playing, looping);
animClip.name           // name of the clip
animClip.track          // track data
animClip.playing        // whether the clip is currently playing
animClip.time           // current clip time
animClip.speed          // play speed of the clip
animClip.loop           // whether the clip loops (either forwards or backwards) or not
animClip.blendWeight    // blend weight for this clip [0..1]
animClip.blendOrder     // order in which this clip should be blended with other clips
animClip.play()         // reset time and enable playing
animClip.stop()         // reset time and disable playing
animClip.pause()        // disable playing
animClip.resume()       // enable playing
animClip.reset()        // reset time

// Setter
AnimSetter = function(value:number[]);

// Target
target = new pc.AnimTarget(func:AnimSetter, type:'vector'|'quaternion', components:number);
target.func             // the setter function which takes as argument the updated animation value
target.type             // specifies the data type for this target
target.components       // the number of components expected for this target (ideally should match the number of components in the attached curves)

// Binder
pc.AnimBinder.joinPath(pathSegments:string[])       // helper function to escape an array of segments and join them into one string
pc.AnimBinder.splitPath(path)                       // helper function to unescape a split a path string into its segments
animBinder.resolve(path)                            // called by the AnimController to resolve a path string to an instance of AnimTarget
animBinder.unresolve(path)                          // called by the AnimController when the path is no longer active (no curves reference this path)
animBinder.update(deltaTime:number)                 // called by the AnimController to give the binder a chance to perform frame update

// Controller
controller = new pc.AnimController(binder);
controller.clips                // the read-only array of clips
controller.addClip(clip)        // add a clip to the controller, new target paths will get resolved by the binder.
controller.removeClip(index)    // remove the clip at the specified index
controller.removeClips()        // remove all clips from the controller
controller.findClip(name)       // returns the first clip with the given name or null if no such clip exists
controller.update(deltaTime)    // update all clips, blend them and set the resulting values on the AnimTargets
```

### CODE REVIEW

File: src/anim/anim.js [hunk 1]
```diff
@@ -508,48 +498,258 @@ Object.assign(pc, function () {
...
-     * @name pc.AnimController
-     * @classdesc AnimContoller stores a set of animation clips and performs blending
-     * between them. It then applies the resulting transforms to the target nodes.
-     * @description Create a new animation controller.
-     * @param {pc.GraphNode} graph - root of the scene graph to control.
+     * @name pc.AnimTarget
+     * @classdesc Stores the information required by {@link pc.AnimController} for updating a target value.
+     * @param {pc.AnimSetter} func - this function will be called when a new animation value is output by
+     * the {@link pc.AnimController}.
+     * @param {'vector'|'quaternion'} type - the type of animation data this target expects.
```
Comment (willeastcott) [inline src/anim/anim.js]: So is a boolean or number target a vector type with 1 component?

**Comment by willeastcott** [comment]
In the PR description, can you summarize the API this PR is intended to introduce (even though it's private for the moment)?

File: src/anim/anim.js (hunk 1 above)
Comment (slimbuck) [reply src/anim/anim.js]: Yes. Presumably a boolean would be a float vec[1] with step interpolation and values limited 0.0's and 1.0's.

**Comment by willeastcott** [comment]
Would it be better to have:

`controller.clips.length` instead of `controller.numClips`
`controller.clips[index]` instead of `controller.getClip(index)`

This would map more closely to:

`entity.children.length`
`entity.chidren[index]`

**Comment by slimbuck** [comment]
AnimController's clips array is not a simple list of clip instances. The controller stores a structure per clip which includes input and output data. Is there a way of providing the interface you suggest without exposing the array directly?

**Comment by willeastcott** [comment]
Could the controller have:

```
controller.clips
controller._inputs // private
controller._outputs // private
```

**Comment by slimbuck** [comment]
Could have, however won't users be tempted to do controller.clips.push()?

**Comment by slimbuck** [comment]
Also, added two missing bits - AnimSetter and AnimTarget

**Comment by willeastcott** [comment]
Users could have the same temptation for `entity.children` but I've never seen anyone make that mistake.

**Comment by slimbuck** [comment]
The AnimController could keep an array of clips just for exposing to clients. Then if a client pushes or modifies the array accidentally AnimController will remain unaffected.

Though this might actually -hide- user bugs and perhaps throwing errors is actually preferable.

**Comment by willeastcott** [comment]
One other comment - are we happy with the naming of `animData.dimension`? Altneratives:

```
animData.dimensions // Plural
animData.components
```

**Comment by slimbuck** [comment]
Exposed AnimController.clips, renamed dimension->dimensions and updated API docs above.

**Comment by willeastcott** [comment]
`AnimData' has `dimensions` and `AnimTarget` has `components`. Is this naming difference justifiable?
